WIP: Panoptes client

eatyourgreens opened the pull request:
Basically removes jQuery and allows for updating client settings via eg. `client.update( {projectToken: 'my new project'})`.

Makes some pretty major changes to the constructor, so this needs testing.

eatyourgreens committed:
Remove jQuery

eatyourgreens committed:
Allow for config updates
Add an update method to handle changes to the client config.
Clean up and simplify the constructor.
Change server option to env for consistency with other apps.

eatyourgreens committed:
Small fixes to Promises

alexbfree commented:
At first look, this looks good. And seems to be working for Panoptes. I/We'll need to check that this is ok in/update Serengeti which I have just remembered is using Geordi client too. https://github.com/zooniverse/Serengeti/blob/master/app/lib/geordi_and_experiments_setup.coffee
Though you won't break anything by updating the client, as it's tied to a specific version. https://github.com/zooniverse/Serengeti/blob/master/package.json#L23

eatyourgreens commented:
I've renamed the `server` config variable to `env`. Beyond that, though, I think the existing client API hasn't changed, so hopefully only small changes would be needed for the Ouroboros apps.

eatyourgreens commented:
Oh, and there may still be a couple of bits where it loads in jQuery Deferred objects from other libs, but treats them as Promises. I've updated those to use `.then()` instead of `.always()`.

alexbfree commented:
at present, Serengeti is the only place Geordi client is used. Galaxy Zoo is still doing it the old (pre-Geordi-client) way. https://github.com/zooniverse/Galaxy-Zoo/blob/adfb9c99f65403ae98e8d07aa1e3ec16b567aff0/app/lib/analytics.coffee

So it should be easy enough for me to update Serengeti. I'll do it when I need to next update Serengeti experiments - no need for you to wait, you can just ignore it for the purposes of what you are doing.

eatyourgreens committed:
Return a promise from logEvent()
Should resolve on success but reject on API or network errors.
Hopefully the API sends back sensible error messages ;)

eatyourgreens commented:
I think this will be good to go once zooniverse/user-string-getter#2 and #9 are merged.

eatyourgreens committed:
update tests

eatyourgreens committed:
Test log requests
Fake Google Analytics dataLayer (otherwise the client dies.)
Add support for XMLHttpRequest in tests.

Commit:
Test logEvent and Update Pass

eatyourgreens committed:
Backward compatibility with older client config

eatyourgreens committed:
Test that server config hasn't broken

eatyourgreens committed:
Update the log test to check the event data

alexbfree commented on src/index.coffee:
```diff
@@ -1,48 +1,52 @@
+    staging: 'https://geordi.staging.zooniverse.org/api/events/'
+    production: 'https://geordi.zooniverse.org/api/events/'
+  
+  GEORDI_DATABASE_FIELDS: [
```
was it deliberate to omit the time fields? Just checking, i think that probably makes sense.

eatyourgreens replied:
I copied the array from this line here https://github.com/zooniverse/geordi-client/blob/master/src/index.coffee#L122
Are there fields missing?

alexbfree replied:
it's ok. found that when i got further down.
I think it was deliberate that the client user shouldn't change those fields

alexbfree commented on src/index.coffee:
```diff
@@ -1,48 +1,52 @@
-    @UserStringGetter = new ZooUserStringGetter(@_getCurrentUserID,@_getCurrentUserIDParameter)
+  zooUserIDGetterParameter: null
+
+  constructor: (config = {}) ->
```
So much nicer! thanks!

alexbfree commented on src/index.coffee:
```diff
@@ -63,48 +67,56 @@ module.exports = class GeordiClient
+            reject Error request.statusText
+
+      request.onerror = () ->
+          reject Error "Network Error"
```
Let's make this a bit more descriptive:
"Network Error while attempting to log with Geordi"
or similar

alexbfree commented on src/index.coffee:
```diff
@@ -119,8 +131,8 @@ module.exports = class GeordiClient
 
   _updateEventDataFromParameterObject: (parameterObject, eventData = {}) ->
     # copy all string & numeric data across
-    for field in ["userID","subjectID","relatedID","errorCode","errorDescription","projectToken","serverURL","experiment","cohort","type","userSeq","sessionNumber","eventNumber","userAgent","clientIP"]
```
OK i see where the list above came from now. Ignore my question about time fields above.
